velox/connectors/hive/storage_adapters/s3fs/tests/S3InsertTest.cpp

```diff
@@ -0,0 +1,176 @@
+    auto hiveConnector =
+        connector::getConnectorFactory(
+            connector::hive::HiveConnectorFactory::kHiveConnectorName)
+            ->newConnector(
+                kHiveConnectorId,
+                minioServer_->hiveConfig(),
+                ioExecutor_.get());
+    connector::registerConnector(hiveConnector);
+  }
+
+  static void TearDownTestSuite() {
+    filesystems::finalizeS3FileSystem();
+    if (minioServer_ != nullptr) {
+      minioServer_->stop();
+      minioServer_ = nullptr;
+    }
+  }
+
+  static std::shared_ptr<MinioServer> minioServer_;
+  static std::unique_ptr<folly::IOThreadPoolExecutor> ioExecutor_;
+};
+std::shared_ptr<MinioServer> S3InsertTest::minioServer_ = nullptr;
+std::unique_ptr<folly::IOThreadPoolExecutor> S3InsertTest::ioExecutor_ =
+    nullptr;
+
+PlanNodePtr createInsertPlan(
+    PlanBuilder& inputPlan,
+    const RowTypePtr& outputRowType,
+    const std::string& outputDirectoryPath,
+    const std::vector<std::string>& partitionedBy = {},
+    const std::shared_ptr<HiveBucketProperty>& bucketProperty = {},
+    const connector::hive::LocationHandle::TableType& outputTableType =
+        connector::hive::LocationHandle::TableType::kNew,
+    const CommitStrategy& outputCommitStrategy = CommitStrategy::kNoCommit) {
+  auto insertTableHandle = std::make_shared<core::InsertTableHandle>(
+      kHiveConnectorId,
+      HiveConnectorTestBase::makeHiveInsertTableHandle(
+          outputRowType->names(),
+          outputRowType->children(),
+          partitionedBy,
+          bucketProperty,
+          HiveConnectorTestBase::makeLocationHandle(
+              outputDirectoryPath, std::nullopt, outputTableType),
+          FileFormat::PARQUET));
+
+  auto insertPlan = inputPlan.tableWrite(
+      inputPlan.planNode()->outputType(),
+      outputRowType->names(),
+      nullptr,
+      insertTableHandle,
+      bucketProperty != nullptr,
+      outputCommitStrategy);
+  return insertPlan.planNode();
+}
+
+TEST_F(S3InsertTest, s3InsertTest) {
+  const int64_t kExpectedRows = 1'000;
+  const char* kOutputDirectory = "s3://writedata/";
+
+  auto rowType = ROW(
+      {"c0", "c1", "c2", "c3"}, {BIGINT(), INTEGER(), SMALLINT(), DOUBLE()});
+
+  auto input = makeRowVector(
+      {makeFlatVector<int64_t>(kExpectedRows, [](auto row) { return row; }),
+       makeFlatVector<int32_t>(kExpectedRows, [](auto row) { return row; }),
+       makeFlatVector<int16_t>(kExpectedRows, [](auto row) { return row; }),
+       makeFlatVector<double>(kExpectedRows, [](auto row) { return row; })});
+
+  minioServer_->addBucket("writedata");
+
+  // Insert into s3 with one writer.
+  auto plan = createInsertPlan(
+      PlanBuilder().values({input}), rowType, kOutputDirectory);
+
+  auto result = AssertQueryBuilder(plan).copyResults(pool());
+
+  auto fragmentVector = result->childAt(TableWriteTraits::kFragmentChannel)
+                            ->asFlatVector<StringView>();
+
+  std::vector<std::string> writeFiles;
```

Comment:
I'm missing some context here. My (maybe n00b) expectation what that it would create a plan with a table write, run it, then check that the correct files were created in the filesystem sink. In that case what does this block of code do?

Maybe related, could use more comments on what this test suite is meant to do.

Reply:
I added some comments. But the TableWrite operator only takes the `targetDirectory` as an argument. It generates the file name. The output of the tableWrite operator contains the information about the filename, row count, etc...

Reply:
I see. And all of that is returned as actual columns in the output dataset.